hail/src/main/scala/is/hail/expr/types/physical/PCanonicalBaseStruct.scala

```diff
@@ -0,0 +1,260 @@
+import is.hail.asm4s.{Code, const, _}
+import is.hail.expr.ir.{EmitMethodBuilder, SortOrder}
+import is.hail.utils._
+
+object PCanonicalBaseStruct {
+  def getByteSizeAndOffsets(types: IndexedSeq[PType], nMissingBytes: Long, byteOffsets: Array[Long]): Long = {
+    assert(byteOffsets.length == types.length)
+    val bp = new BytePacker()
+
+    var offset: Long = nMissingBytes
+    types.zipWithIndex.foreach { case (t, i) =>
+      val fSize = t.byteSize
+      val fAlignment = t.alignment
+
+      bp.getSpace(fSize, fAlignment) match {
+        case Some(start) =>
+          byteOffsets(i) = start
+        case None =>
+          val mod = offset % fAlignment
+          if (mod != 0) {
+            val shift = fAlignment - mod
+            bp.insertSpace(shift, offset)
+            offset += (fAlignment - mod)
+          }
+          byteOffsets(i) = offset
+          offset += fSize
+      }
+    }
+    offset
+  }
+
+  def alignment(types: Array[PType]): Long = {
+    if (types.isEmpty)
+      1
+    else
+      types.map(_.alignment).max
+  }
+}
+
+trait PCanonicalBaseStruct extends PBaseStruct {
+  lazy val fieldRequired: Array[Boolean] = types.map(_.required)
+
+  lazy val fieldIdx: Map[String, Int] =
+    fields.map(f => (f.name, f.index)).toMap
+
+  lazy val fieldNames: Array[String] = fields.map(_.name).toArray
+
+  def fieldByName(name: String): PField = fields(fieldIdx(name))
+
+  def index(str: String): Option[Int] = fieldIdx.get(str)
+
+  def selfField(name: String): Option[PField] = fieldIdx.get(name).map(i => fields(i))
+
+  def hasField(name: String): Boolean = fieldIdx.contains(name)
+
+  def field(name: String): PField = fields(fieldIdx(name))
+
+  def fieldType(name: String): PType = types(fieldIdx(name))
+
+  lazy val size: Int = fields.length
+
+  def _toPretty: String = {
+    val sb = new StringBuilder
+    _pretty(sb, 0, compact = true)
+    sb.result()
+  }
+
+  def identBase: String
+  def _asIdent: String = {
+    val sb = new StringBuilder
+    sb.append(identBase)
+    sb.append("_of_")
+    types.foreachBetween { ty =>
+      sb.append(ty.asIdent)
+    } {
+      sb.append("AND")
+    }
+    sb.append("END")
+    sb.result()
+  }
+
+  def codeOrdering(mb: EmitMethodBuilder, so: Array[SortOrder]): CodeOrdering =
+    codeOrdering(mb, this, so)
+
+  def codeOrdering(mb: EmitMethodBuilder, other: PType, so: Array[SortOrder]): CodeOrdering
+
+  def isIsomorphicTo(other: PBaseStruct): Boolean =
+    size == other.size && isCompatibleWith(other)
+
+  def isPrefixOf(other: PBaseStruct): Boolean =
+    size <= other.size && isCompatibleWith(other)
+
+  def isCompatibleWith(other: PBaseStruct): Boolean =
+    fields.zip(other.fields).forall{ case (l, r) => l.typ isOfType r.typ }
+
+  def truncate(newSize: Int): PBaseStruct
+
+  override def unsafeOrdering(): UnsafeOrdering =
+    unsafeOrdering(this)
+
+  override def unsafeOrdering(rightType: PType): UnsafeOrdering = {
```

Comment:
this should be defined on the abstract superclass.

Reply:
In the PArray PR, we put this on the mixed in implementation to which this mixin is an analog: https://github.com/hail-is/hail/pull/7687/files#diff-1c9014eb90e31ec94b9c095c9ec0a9abR165

Reply:
yeah, that's wrong. Let's fix that too. The right design requires multiple iterations :)

The reason this is wrong is that it's implemented in terms of the PContainer interface methods.

Reply:
(so it can happily be defined on the interface)

Reply:
> yeah, that's wrong. Let's fix that too

sure thing!

Reply:
https://github.com/hail-is/hail/issues/7737